### PR TITLE
Enumerables with Ruby

### PR DESCRIPTION
**Pull request by Abidoyinsola1**
- add setup project
- add git-flow
- add the methods and module

### CODE REVIEW

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi  Team 👋

Nice work on this so far. 👍

However, let's fix these issues:

- [x] The following should return an enumerator when no block is given:
- my_each
- my_each_with_index
- my_select
- my_map

- [x] my_each_with_index throws an error when a range, array, or hash is passed in an example like this `(1..20).my_each_with_index(&block)`.

- [x] my_all fails in these cases: `[1,true,2,true].my_all`, `[null, false, false, null].my_all`. It also fails when a Regex, Ruby class or pattern(`[1,2,3].my_all(3)`) is passed.

- [x] my_any, my_none, my_map, my_inject, and multiply_els also don't work as expected.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
+module Enumerable
+  def my_each
+    return unless block_given?
+
+    each do |i|
+      yield i
+    end
+    self
+  end
```
Comment (christianotieno) [inline main.rb]: # my_each

- [x] `my_each` returns an Enumerator if no block is given
     Failure/Error: expect(array.my_each).to be_an(Enumerator)

       expected nil to be a kind of Enumerator

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_each_with_index
+    return unless block_given?
+
+    0.upto(length - 1) do |i|
+      yield i, index
+      index + 1
+    end
+  end
```
Comment (christianotieno) [inline main.rb]: # my_each_with_index
- [x] `my_each_with_index` returns an enumerator if no block is given

     Failure/Error: expect(array.my_each_with_index).to be_an(Enumerator)
       expected nil to be a kind of Enumerator
##

- [x] `my_each_with_index` when a block is given when self is an array returns the array itself after calling the given block once for each element in self

     Failure/Error: expect(array.my_each_with_index(&block)).to eq(array.each_with_index(&block))

       expected: [4, 2, 6, 0, 2, 7, 5, 3, 3, 1, 2, 1, 3, 8, 8, 1, 2, 7, 5, 3, 7, 8, 1, 3, 1, 5, 4, 0, 4, 8, 2, 7, 0, 0..., 8, 1, 7, 3, 5, 5, 6, 7, 5, 7, 4, 0, 2, 6, 2, 2, 7, 8, 6, 6, 4, 1, 1, 7, 7, 4, 1, 1, 5, 7, 4, 2, 3]
            got: 0

##

- [x] `my_each_with_index` when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self

     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)

       expected: "Num: 5, idx: 0\nNum: 5, idx: 1\nNum: 5, idx: 2\nNum: 1, idx: 3\nNum: 1, idx: 4\nNum: 3, idx: 5\nNum:...: 4, idx: 94\nNum: 5, idx: 95\nNum: 4, idx: 96\nNum: 5, idx: 97\nNum: 8, idx: 98\nNum: 1, idx: 99\n"
            got: "Num: 0, idx: #<Enumerator:0x0000559ab87450d8>\nNum: 1, idx: #<Enumerator:0x0000559ab8744f48>\nNum: 2...>\nNum: 98, idx: #<Enumerator:0x0000559ab8773280>\nNum: 99, idx: #<Enumerator:0x0000559ab8773118>\n"

##

- [x] `my_each_with_index` when a block is given when self is a range returns the range itself after calling the given block once for each element in self

     Failure/Error: expect(range.my_each_with_index(&block)).to eq(range.each_with_index(&block))

     NameError:
       undefined local variable or method `length' for 5..50:Range
##

- [x] `my_each_with_index` when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

     Failure/Error: enum.my_each_with_index(&block)

     NameError:
       undefined local variable or method `length' for 5..50:Range

##

- [x] `my_each_with_index` when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

     Failure/Error: enum.my_each_with_index(&block)

     ArgumentError:
       wrong number of arguments (given 0, expected 1)
##

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_select
+    return unless block_given?
+
+    result = []
+    my_each do |i|
+      result.push(i) if yield i
+    end
+    result
+  end
```
Comment (christianotieno) [inline main.rb]: # my_select

- [x] `my_select` returns an enumerator if no block is given

     Failure/Error: expect(array.my_select).to be_an(Enumerator)
       expected nil to be a kind of Enumerator

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    result = true
+    if !val[0].nil?
+      my_each { |i| result = false unless val[0] == i }
+    elsif !block_given?
+      my_each { |i| result = false unless i }
+    else
+      my_each { |i| result = false unless yield(i) }
+    end
+    result
+  end
```
Comment (christianotieno) [inline main.rb:39]: # my_all?  
- [x] `my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

     Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)

       expected true
            got false

- [x] `my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric

      Failure/Error: expect(numbers.my_all?(Numeric)).to be numbers.all?(Numeric)

        expected true
             got false

- [x] `my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex

      Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)

        expected true
             got false

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
```
Comment (christianotieno) [inline main.rb]: # my_any?
- [x] `my_any?` returns false if the block ever returns a value false or nil::range

      Failure/Error: expect(range.my_any?(&false_block)).to eq(range.any?(&false_block))

        expected: true
             got: false

##
- [x] `my_any?` when no block or argument is given returns true if at least one of the collection is not false or nil

      Failure/Error: expect(true_array.my_any?).to be true_array.any?

        expected true
             got false
 ##
- [x] `my_any?` when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

      Failure/Error: expect(array.my_any?(Numeric)).to be array.any?(Numeric)

        expected true
             got false
##      
- [x] `my_any?` when a Regex is passed as an argument returns true if any of the collection matches the Regex

      Failure/Error: expect(words.my_any?(/d/)).to be words.any?(/d/)

        expected true
             got false
##      
- [x] `my_any?` when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

      Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')

        expected true
             got false

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    result = false
+    if !val[0].nil?
+      my_each { |i| result = false unless val[0] == i }
+    elsif !block_given?
+      my_each { |i| result = false unless i }
+    else
+      my_each { |i| result = false if yield(i) }
+    end
+    result
+  end
```
Comment (christianotieno) [inline main.rb]: # my_none?
- [x] `my_none?` returns true if the block never returns true for all elements

      Failure/Error: expect(array.my_none?(&true_block)).to eq(array.none?(&true_block))

        expected: true
             got: false
##
- [x] `my_none?` when no block or argument is given returns true only if none of the collection members is true

      Failure/Error: expect(false_array.my_none?).to be true

        expected true
             got false
##
- [x] `my_none?` when a Regex is passed as an argument returns true if none of the collection matches the Regex

      Failure/Error: expect(words.my_none?(/z/)).to be words.none?(/z/)

        expected true
             got false
     
- [x] `my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

      Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

        expected true
             got false

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_count(*)
+    return length unless block_given?
+
+    count = 0
+    my_each do |i|
+      count += 1 if yield i
+    end
+    count
+  end
```
Comment (christianotieno) [inline main.rb]: # my_count
- [x] `my_count` returns the number of items in enum through enumeration::range

      Failure/Error: expect(range.my_count).to eq range.count

      NameError:
        undefined local variable or method `length' for 5..50:Range
##
- [x] `my_count` counts the number of items in enum that are equal to an item if an argument is given

      Failure/Error: expect(array.my_count(LOWEST_VALUE)).to eq array.count(LOWEST_VALUE)

        expected: 12
             got: 100

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      end
+    elsif !sym.nil? && sym.class == String && %r{[+-/*]}.match(sym)
+      while i < length
+        acc = acc.send(sym, self[i])
+        i += 1
+      end
+    end
+    acc
+  end
+end
```
Comment (christianotieno) [inline main.rb]: # my_inject
- [ ] `my_inject` raises a "LocalJumpError" when no block or argument is given

      Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
        expected LocalJumpError but nothing was raised
##
- [x] `my_inject` when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range

      Failure/Error: actual = range.my_inject { |prod, n| prod * n }

      NameError:
        undefined local variable or method `length' for 5..50:Range
##   
- [x] `my_inject` when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block

      Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }

      NameError:
        undefined local variable or method `length' for 5..50:Range
##
- [x] `my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method

      Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)

      NoMethodError:
        undefined method `+' for :+:Symbol
##
- [x] `my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range

      Failure/Error: expect(range.my_inject(:*)).to eq range.inject(:*)

      NameError:
        undefined local variable or method `length' for 5..50:Range
##
- [ ] `my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method

      Failure/Error: expect(array.my_inject(20, :*)).to eq array.inject(20, :*)

      ArgumentError:
        wrong number of arguments (given 2, expected 0..1)
##
- [ ] `my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

      Failure/Error: expect(range.my_inject(2, :*)).to eq range.inject(2, :*)

      ArgumentError:
        wrong number of arguments (given 2, expected 0..1)

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey team :wave:
-
You have done a splendid job with the changes! :tada: Keep it up.
We, however, do have some few more changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Tests Done: 79
## Failures: 29
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### Well done, you got it right this time :clap: and the project has been approved. Congratulations :boom: :100:
### You can merge and submit the project. Good luck with the next project.
- Happy coding!

**Pull request merged**